File: lib/edge/python/examples/bm25-search.py
"""Native BM25 sparse search: embed text manually with `Bm25`, then upsert and query
the produced sparse vectors through the regular update/query path.
"""

import os, shutil
from pathlib import Path

from qdrant_edge import (
    EdgeShard, EdgeConfig, EdgeSparseVectorParams, Modifier,
    Bm25, Bm25Config,
    Point, Query, QueryRequest, UpdateOperation,
)


DATA_DIR = Path(__file__).parent.parent.parent / "data"
TMP_DIR = DATA_DIR / "tmp"

path = TMP_DIR / "qdrant_edge_bm25"
shutil.rmtree(path, ignore_errors=True)
os.makedirs(path)

config = EdgeConfig(
    sparse_vectors={"text": EdgeSparseVectorParams(modifier=Modifier.Idf)},
)
shard = EdgeShard.create(str(path), config)

bm25 = Bm25(Bm25Config(language="english"))

shard.update(UpdateOperation.upsert_points([
    Point(1, {"text": bm25.embed_document("the quick brown fox")}, {"doc": "1"}),
    Point(2, {"text": bm25.embed_document("a lazy dog sleeps")},   {"doc": "2"}),
    Point(3, {"text": bm25.embed_document("foxes are clever")},    {"doc": "3"}),
]))
shard.optimize()
print(f"Info: {shard.info()}")

q = bm25.embed_query("clever fox")
results = shard.query(QueryRequest(
    limit=3, query=Query.Nearest(q, using="text"), with_payload=True,
))
print(f"Got {len(results)} results")
for r in results:
    print(f"  id={r.id} score={r.score:.4f} payload={r.payload}")

File: lib/edge/publish/examples/src/bin/bm25-search.rs
// See lib/edge/python/examples/bm25-search.py for the equivalent Python example.
//
// Native BM25 sparse search: embed text with `EdgeBm25`, then upsert and query
// the produced sparse vectors through the regular update/query path.

use std::collections::HashMap;
use std::error::Error;
use std::path::Path;

use examples::TMP_DIR;
use qdrant_edge::bm25_embed::{EdgeBm25, EdgeBm25Config};
use qdrant_edge::external::serde_json::json;
use qdrant_edge::{
    EdgeConfig, EdgeShard, EdgeSparseVectorParams, Modifier, NamedQuery, PointInsertOperations,
    PointOperations, PointStruct, QueryEnum, QueryRequest, ScoringQuery, UpdateOperation,
    VectorInternal, Vectors, WithPayloadInterface, WithVector,
};

const SPARSE_VECTOR_NAME: &str = "text";

fn main() -> Result<(), Box<dyn Error>> {
    let path = Path::new(TMP_DIR).join("qdrant_edge_bm25");
    if path.exists() {
        fs_err::remove_dir_all(&path)?;
    }
    fs_err::create_dir_all(&path)?;

    let config = EdgeConfig {
        sparse_vectors: HashMap::from([(
            SPARSE_VECTOR_NAME.to_string(),
            EdgeSparseVectorParams {
                modifier: Some(Modifier::Idf),
                ..Default::default()
            },
        )]),
        ..Default::default()
    };
    let shard = EdgeShard::new(&path, config)?;

    let bm25 = EdgeBm25::new(EdgeBm25Config {
        language: Some("english".to_string()),
        ..Default::default()
    })?;

    let docs = [
        (1u64, "the quick brown fox", "1"),
        (2, "a lazy dog sleeps", "2"),
        (3, "foxes are clever", "3"),
    ];

    let points = docs
        .iter()
        .map(|(id, text, doc)| {
            PointStruct::new(
                *id,
                Vectors::new_named([(SPARSE_VECTOR_NAME, bm25.embed_document(text))]),
                json!({ "doc": doc }),
            )
            .into()
        })
        .collect();

    shard.update(UpdateOperation::PointOperation(
        PointOperations::UpsertPoints(PointInsertOperations::PointsList(points)),
    ))?;
    shard.optimize()?;
    println!("Info: {:?}", shard.info());

    let query = bm25.embed_query("clever fox");
    let results = shard.query(QueryRequest {
        prefetches: vec![],
        query: Some(ScoringQuery::Vector(QueryEnum::Nearest(NamedQuery {
            query: VectorInternal::from(query),
            using: Some(SPARSE_VECTOR_NAME.to_string()),
        }))),
        filter: None,
        score_threshold: None,
        limit: 3,
        offset: 0,
        params: None,
        with_vector: WithVector::Bool(false),
        with_payload: WithPayloadInterface::Bool(true),
    })?;

    println!("Got {} results", results.len());
    for r in &results {
        println!(
            "  id={:?} score={:.4} payload={:?}",
            r.id, r.score, r.payload
        );
    }

    Ok(())
}
